File: solutions/csharp/grade-school/1/GradeSchool.cs
public class GradeSchool
{
    private HashSet<string> students = new();
    private Dictionary<int, List<string>> grades = new();
    public bool Add(string student, int grade)
    {
        if (students.Add(student))
        {
            if (!grades.ContainsKey(grade))
            {
                grades[grade] = new List<string>();
                grades[grade].Add(student);
            }
            else
            {
                grades[grade].Add(student);
            }
            return true;
        }
        return false;
    }

    public IEnumerable<string> Roster()
    {
        return grades.OrderBy(g => g.Key)
                      .SelectMany(g => g.Value.OrderBy(s => s));
    }

    public IEnumerable<string> Grade(int grade)
    {
        if(grades.ContainsKey(grade))
        {
            return grades[grade].OrderBy(s => s);
        }
        return Enumerable.Empty<string>();
    }
}

File: solutions/cpp/grade-school/1/grade_school.cpp
#include "grade_school.h"

namespace grade_school {
    const std::map<int, std::vector<std::string>>& school::roster() const{
        return this->current_roster;
    }

    bool school::empty(){
        return this->current_roster.empty();
    }
    
    void school::add(const std::string& name, const int grade){
        if(this->current_roster.count(grade)){
            this->current_roster[grade].emplace_back(name);
            std::sort(this->current_roster[grade].begin(), this->current_roster[grade].end());
        }
        else{
            this->current_roster[grade] = std::vector<std::string>{name};
        }
    }

    std::vector<std::string> school::grade(int grade) const {
    auto it = current_roster.find(grade);
    if (it != current_roster.end()) {
        auto sorted_list = it->second;
        return sorted_list;
    }
    return {};
}
}  // namespace grade_school
